### PR TITLE
Blog App: validations and model specs

### PR DESCRIPTION
**Pull request by Amazinggracee**
### Project requirements

- [ ] Add the following validations:
      - For the User model:
        1. Name must not be blank.
        2. PostsCounter must be an integer greater than or equal to zero.
    - For the Post model:
      1. Title must not be blank.
      2. Title must not exceed 250 characters.
      3. CommentsCounter must be an integer greater than or equal to zero.
      4. LikesCounter must be an integer greater than or equal to zero.
- [ ] Add unit specs for all of your models' methods and validations.

### CODE REVIEW

File: spec/models/user_spec.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+  it 'is not valid with a non-integer posts_counter' do
+    user = User.new(name: 'John Doe', posts_counter: 5.5)
+    expect(user).not_to be_valid
+  end
+
+  it 'is valid with a zero posts_counter' do
+    user = User.new(name: 'John Doe', posts_counter: 0)
+    expect(user).to be_valid
+  end
```
Comment (danmainah) [inline spec/models/user_spec.rb:27]: - [ ] @Amazinggracee Great work adding tests for your validation, however, you forgot to add tests for the custom method `recent_posts`, it's part of the project requirement
image

File: spec/models/post_spec.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+      post = user.posts.new(title: 'Sample Title', content: 'Sample Content', comments_counter: 0, likes_counter: 5.5)
+      expect(post).not_to be_valid
+    end
+
+    it 'is not valid without content' do
+      user = User.create(name: 'John Doe')
+      post = user.posts.new(title: 'Sample Title', content: nil, comments_counter: 0, likes_counter: 0)
+      expect(post).not_to be_valid
+    end
+  end
```
Comment (danmainah) [inline spec/models/post_spec.rb:78]: - [ ] @Amazinggracee Great work adding tests for your validation, however, you forgot to add tests for the custom method `recent_comments`, it's part of the project requirement

 - [ ] @Amazinggracee You forgot to add tests for your like and comment models, kindly consider adding tests for them.
 
image

**Review by danmainah — CHANGES_REQUESTED** [review]
Hi @Amazinggracee , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
✔️ Added validation for the user and post models
✔️ Ensured all added tests are passing.


# Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: spec/models/user_spec.rb (hunk 1 above)
Comment (Amazinggracee) [reply spec/models/user_spec.rb:27]: My apologies, I have added the test. Thank you, Dan.

File: app/models/post.rb [hunk 3]
```diff
@@ -1,14 +1,20 @@
...
+  after_save :count_posts
+
+  validates :title, presence: true, length: { in: 1..250 }
+  validates :comments_counter, numericality: { only_integer: true, greater_than_or_equal_to: 0 }
+  validates :likes_counter, numericality: { only_integer: true, greater_than_or_equal_to: 0 }
 
-  def recent_comments
+  def most_recent_comments
     comments.order(created_at: :desc).limit(5)
   end
```
Comment (elisha2kyakpo1) [inline app/models/post.rb:13]: - [ ] Kindly consider testing this method as requested by the previous reviewer. We want to know if the 5 most recent comments are being returned on each post.

File: spec/models/user_spec.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+      expect(user.recent_posts).to eq([post1, post2])
+    end
+
+    it 'returns only the specified number of recent posts' do
+      user = User.create(name: 'Eve White', posts_counter: 4)
+      posts = []
+      5.times { posts << user.posts.create(title: 'Post Title', content: 'Post Content') }
+
+      expect(user.recent_posts(limit: 3).count).to eq(3)
```
Comment (elisha2kyakpo1) [inline spec/models/user_spec.rb:56]: - [ ] Kindly make sure all the tests are passing before you submit again for review. I have noticed that about 4 tests are failing.


test

**Review by elisha2kyakpo1 — CHANGES_REQUESTED** [review]
Hi 👋 @Amazinggracee, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights

✔️ The PR title and description are good and professional.
✔️ Recent posts method tested.
✔️ Good communication with code reviewer👏
👏 The app works as expected.

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/models/user_spec.rb (hunk 4 above)
Comment (Amazinggracee) [reply spec/models/user_spec.rb:56]: Hello @elisha2kyakpo1 , how are you doing today? The 21 test I wrote Passed, here is the screen shot. 
image
 I had to involve some of my friends who cloned my RP just to test the app.

File: app/models/post.rb (hunk 3 above)
Comment (Amazinggracee) [reply app/models/post.rb:13]: I also added the test.

**Review by BertrandConxy — APPROVED** [review]
Hi @Amazinggracee, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights 💪 
- No linter errors
- project requirements are met
- project works fine

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @BertrandConxy in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**